File: twister-lib-android/slf4j/src/main/java/net/twisterrob/android/log/AndroidLoggerServiceProvider.java
package net.twisterrob.android.log;

import org.slf4j.ILoggerFactory;
import org.slf4j.IMarkerFactory;
import org.slf4j.helpers.BasicMarkerFactory;
import org.slf4j.helpers.NOPMDCAdapter;
import org.slf4j.spi.MDCAdapter;
import org.slf4j.spi.SLF4JServiceProvider;

public class AndroidLoggerServiceProvider implements SLF4JServiceProvider {

	/**
	 * Declare the version of the SLF4J API this implementation is compiled against. 
	 * The value of this field is modified with each major release. 
	 */
	// to avoid constant folding by the compiler, this field must *not* be final
	public static String REQUESTED_API_VERSION = "2.0.9"; // !final

	private ILoggerFactory loggerFactory;
	private IMarkerFactory markerFactory;
	private MDCAdapter mdcAdapter;

	@Override
	public void initialize() {
		loggerFactory = new AndroidLoggerFactory();
		markerFactory = new BasicMarkerFactory();
		mdcAdapter = new NOPMDCAdapter();
	}

	public ILoggerFactory getLoggerFactory() {
		return loggerFactory;
	}

	@Override
	public IMarkerFactory getMarkerFactory() {
		return markerFactory;
	}

	@Override
	public MDCAdapter getMDCAdapter() {
		return mdcAdapter;
	}

	@Override
	public String getRequestedApiVersion() {
		return REQUESTED_API_VERSION;
	}
}

File: twister-lib-android/slf4j/src/main/java/net/twisterrob/android/log/AndroidLoggerServiceProvider.kt
package net.twisterrob.android.log

import org.slf4j.ILoggerFactory
import org.slf4j.IMarkerFactory
import org.slf4j.helpers.BasicMarkerFactory
import org.slf4j.helpers.NOPMDCAdapter
import org.slf4j.spi.MDCAdapter
import org.slf4j.spi.SLF4JServiceProvider

// TODO revert to earlier state with val/vars https://youtrack.jetbrains.com/issue/KT-6653#focus=Comments-27-9920359.0-0
class AndroidLoggerServiceProvider : SLF4JServiceProvider {

	private val requestedApiVersion: String = "2.0.13"
	override fun getRequestedApiVersion(): String =
		requestedApiVersion

	private lateinit var loggerFactory: ILoggerFactory
	override fun getLoggerFactory(): ILoggerFactory =
		loggerFactory

	private val markerFactory: IMarkerFactory = BasicMarkerFactory()
	override fun getMarkerFactory(): IMarkerFactory =
		markerFactory

	private val mdcAdapter: MDCAdapter = NOPMDCAdapter()
	override fun getMDCAdapter(): MDCAdapter =
		mdcAdapter

	override fun initialize() {
		loggerFactory = AndroidLoggerFactory()
	}
}
